sdks/java/core/src/test/java/org/apache/beam/sdk/metrics/MetricsTest.java

```diff
@@ -69,216 +74,286 @@ private static MetricQueryResults queryTestMetrics(PipelineResult result) {
-    Metrics.distribution("", NAME);
+  /** Tests for committed metrics. */
+  @RunWith(JUnit4.class)
+  public static class CommittedMetricTests extends SharedTestBase {
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesCounterMetrics.class,
+        UsesDistributionMetrics.class, UsesGaugeMetrics.class})
+    @Test
+    public void testAllCommittedMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+
+      assertAllMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesCounterMetrics.class})
+    @Test
+    public void testCommittedCounterMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertCounterMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesDistributionMetrics.class})
+    @Test
+    public void testCommittedDistributionMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertDistributionMetrics(metrics, true);
+    }
+
+    @Category({ValidatesRunner.class, UsesCommittedMetrics.class, UsesGaugeMetrics.class})
+    @Test
+    public void testCommittedGaugeMetrics() {
+      PipelineResult result = runPipelineWithMetrics();
+      MetricQueryResults metrics = queryTestMetrics(result);
+      assertGaugeMetrics(metrics, true);
+    }
+
+    @Test
+    @Category({NeedsRunner.class, UsesAttemptedMetrics.class, UsesCounterMetrics.class})
+    public void testBoundedSourceMetrics() {
+      long numElements = 1000;
+
+      pipeline.apply(GenerateSequence.from(0).to(numElements));
+
+      PipelineResult pipelineResult = pipeline.run();
+
+      MetricQueryResults metrics =
+          pipelineResult
+              .metrics()
+              .queryMetrics(
+                  MetricsFilter.builder()
+                      .addNameFilter(
+                          MetricNameFilter.named(
+                              ELEMENTS_READ.getNamespace(), ELEMENTS_READ.getName()))
+                      .build());
+
+      assertThat(metrics.getCounters(), hasItem(
+          attemptedMetricsResult(
+              ELEMENTS_READ.getNamespace(),
+              ELEMENTS_READ.getName(),
+              "Read(BoundedCountingSource)",
+              1000L)));
+    }
+
+    @Test
+    @Category({NeedsRunner.class, UsesAttemptedMetrics.class, UsesCounterMetrics.class})
+    public void testUnboundedSourceMetrics() {
+      long numElements = 1000;
+
+      // Use withMaxReadTime to force unbounded mode.
+      pipeline.apply(
+          GenerateSequence.from(0).to(numElements).withMaxReadTime(Duration.standardDays(1)));
+
+      PipelineResult pipelineResult = pipeline.run();
+
+      MetricQueryResults metrics =
+          pipelineResult
+              .metrics()
```

Comment:
same as above. Actually, I think this is true for all the tests using `PepelineResults.metrics()`

Reply:
See comment above. If the tests are exercising different behaviors of `PipelineResults.metrics()` that are runner-provided functionality, we should keep them all as `@ValidatesRunner`. Otherwise, having at least one as `@ValidatesRunner` is sufficient and the rest can be `@NeedsRunner`